beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -376,3 +375,77 @@ func (f *blocksFetcher) requestBlocks(
 	}
 	return blocks, nil
 }
+
+// requestBlocksByRoot is a wrapper for handling BeaconBlockByRootsReq requests/streams.
+func (f *blocksFetcher) requestBlocksByRoot(
+	ctx context.Context,
+	req *p2pTypes.BeaconBlockByRootsReq,
+	pid peer.ID,
+) ([]*eth.SignedBeaconBlock, error) {
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+	if req == nil {
+		return nil, errors.New("nil request")
+	}
+	l := f.getPeerLock(pid)
+	if l == nil {
+		return nil, errors.New("cannot obtain lock")
+	}
+	l.Lock()
+	log.WithFields(logrus.Fields{
+		"peer":     pid,
+		"numRoots": len(*req),
+		"capacity": f.rateLimiter.Remaining(pid.String()),
+		"score":    f.p2p.Peers().Scorers().BlockProviderScorer().FormatScorePretty(pid),
+	}).Debug("Requesting blocks (by roots)")
+	if f.rateLimiter.Remaining(pid.String()) < int64(len(*req)) {
+		if err := f.waitForBandwidth(pid); err != nil {
+			return nil, err
+		}
+	}
+	f.rateLimiter.Add(pid.String(), int64(len(*req)))
+	l.Unlock()
+
+	stream, err := f.p2p.Send(ctx, req, p2p.RPCBlocksByRootTopic, pid)
```

Comment:
this whole logic is repeated. We do the same exact things in the sync package, is there are any way to abstract this rather than fully copying it ?

Reply:
I think we should refactor this into its own helper

Reply:
Ok, will do. Thank you for suggestion.

Reply:
Done in #7718